### PR TITLE
radio rework

### PR DESCRIPTION
**Pull request by deltanedas**
## About the PR
reworks radio to support non-prototype channels
intended to make better handheld radio a lot simpler and cleaner

## Why / Balance
see handheld pr

## Technical details
- RadioSystem no longer uses channel prototypes, it uses frequencies for checking if you can hear radio messages etc
- you can now create a custom RadioChannel and use it instead of a prototype, for example to set the channel's frequency to user input.
- various radio systems get frequencies from prototypes

## Media
headset works

handhelds work

intercom mic works

intercom speaker works

borg on intercom

some machines

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
- radio transmission is based on frequencies rather than channel prototype ids now, as non-prototype channels are supported.
- `ActiveRadioComponent` now uses access, you must use `SetChannels`/`SetFrequencies`/`Disable` instead of setting channels the old way.
- radio events contain `RadioChannel`s now instead of prototypes, attempt to cast them if you really need the prototype id.

**Changelog**
no cl no fun

### CODE REVIEW

**Comment by LordEclipse** [comment]
holy based

**Comment by QuietlyWhisper** [comment]
How does this interact with telecomms?

**Comment by CatMagic1** [comment]
I was thinking of this earlier and it seems I willed it into existence.

Epic delta

**Comment by deltanedas** [comment]
> How does this interact with telecomms?

unanchor server and comms goes down same as before

for custom radio channels long range will ignore comms server

**Comment by Arteben** [comment]
This looks a bit complicated. We're introducing a second entity that's somewhat similar to channels, and it's easy to get confused between these channels and their frequencies.

Perhaps we can take a different approach. Expand the frequency field in the radio channels component, allowing it to specify not just one frequency, but a range of frequencies. Additionally, add a field for the active frequency, which by default will be 0 or the first frequency for that channel. The field can be used for handled radiostations.

**Comment by deltanedas** [comment]
i dunno if using a range would work for some places where frequencies get used

**Comment by Arteben** [comment]
For devices, use a method with the parameter of the active frequency, which defaults to 0. This parameter can also be specified in the prototypes for channels. Two devices that use different frequencies of the same channel cannot communicate with each other but use the same name and tags for chat. Events that disable channels should also disable all their frequencies.

Your solution works(i did a little tests), but it was challenging to understand the application of frequencies and channels. And my eye twitches when I see two very similar methods that differ only in one parameter, example, SetChannel & SetFrequency.

**Comment by deltanedas** [comment]
theyre to avoid unnecessary computation since the frequency is already known but channel needs to look it up from prototype, i think

File: Content.Server/Radio/EntitySystems/RadioDeviceSystem.cs [hunk 1]
```diff
@@ -206,6 +207,15 @@ private void OnReceiveRadio(EntityUid uid, RadioSpeakerComponent component, ref
...
     }
 
+    private void OnIntercomInit(Entity<IntercomComponent> ent, ref MapInitEvent args)
+    {
+        foreach (var id in ent.Comp.SupportedChannels)
+        {
+            var channel = _protoMan.Index<RadioChannelPrototype>(id);
+            ent.Comp.SupportedFrequencies.Add(channel.Frequency);
+        }
+    }
```
Comment (mirrorcult) [inline Content.Server/Radio/EntitySystems/RadioDeviceSystem.cs:217]: i dont really like this pattern of having to separately store frequencies and bookkeep it to make sure all the loaded prototypes get their frequencies added. an alternative is to maybe make frequency just into a string and have it all in one list. try to index it as a radio channel first and if it doesnt work assume its a frequency? maybe janky

**Comment by deltanedas** [comment]
will just close since too sus

**Pull request closed without merging**